rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+
+Extension points will be defined and exposed via PF4J.
+
+### PF4J as a Foundation
+[PF4J](https://github.com/pf4j/pf4j) is to be used as the foundation for all plugin functionality. While designed for in-process extensibility, we’ll be using PF4J to define both the contract for in-process plugins, as well as remote plugins. Remote plugins would be supported in two parts: First, by including a plugin that activates a new plugin runtime (e.g. gRPC, or a Docker co-process). Second, a remote plugin stub that uses the runtime plugin as a transport.
+
+PF4J has a lot to offer out of the box that we would need to build ourselves, regardless of final implementation. It doesn’t go as far as OSGI, but provides basic (yet extensible) ClassLoader isolation and a simple extension contract that we can use broadly. PF4J was originally built for performing in-process plugins only, but its own extension points make it feasible to enable remote plugins using the same contract. What runtime a particular plugin uses would be determined by a combination of: 1) The extension point developer allowing remote invocation, and 2) an operator selecting the particular runtime they want to use from the available options.
+
+An extension looks like the following:
+
+```kotlin
+// Definition of the PluginStage extension point.
+// The annotation defined here also informs the Spinnaker
+// plugin framwework which runtimes are acceptable. This 
+// annotation would additionally be used for other metadata
+// Spinnaker needs to know atop whatever PF4J offers out of
+// the box.
+@SpinnakerExtension(drivers = [
+  ExtensionDriver.LOCAL, 
+	ExtensionDriver.REMOTE
+])
+class PipelineStageExtension : ExtensionPoint {
+	// ...
+}
+```
+
+It is highly recommended to read the documentation of PF4J to fully understand the scope of this proposal.
+
+### Front50 as the Source of Truth
+Operators will need the ability to understand, at a glance, what plugins are installed and enabled or disabled, their versions, as well as any other metadata about a plugin. Whether plugins are implemented as in-process or not, they Front50 should be used to track and expose a single source of truth of what plugins are installed in any particular service, as well as which ones are enabled or disabled. 
+
+The Front50 plugin registry will also be exposed through Gate, and can be consumed by Deck to load the correct plugins on application startup.
+
+Plugins can also be enabled or disabled at runtime without redeploying a service. This flag is sourced from Front50, which all services will be listening for change events from. Whether this change event is published or polled for initially is unknown.
+
+Halyard configs can be used to inform services of their _initial state_. When a service starts up, it will notify Front50 of what in-process plugins are installed, however Front50 will be responsible for determining the enabled state.
```

Comment:
I will update. Here's the difference:

A plugin is "installed" if it's in the system. An installed plugin can either be in an "enabled" or "disabled" state. In a disabled state, it will not be getting invoked and will not have its internal components started. If the plugin is transitioning state from enabled to disabled (or vice-versa), the plugin framework calls a correlated method to handle whatever logic the plugin must step through to become running or stopped.

I will update the doc to cover this lifecycle. It's dictated (mostly) by PF4J but will have some deviations based on our decision to use Front50 as SOT.